components/region_cache_memory_engine/src/write_batch.rs

```diff
@@ -26,6 +28,7 @@ pub(crate) const NODE_OVERHEAD_SIZE_EXPECTATION: usize = 96;
 // As every key/value holds a Arc<MemoryController>, this overhead should be
 // taken into consideration.
 pub(crate) const MEM_CONTROLLER_OVERHEAD: usize = 8;
+const AMOUNT_TO_CLEAN_TOMBSTONE: u64 = ReadableSize::mb(16).0;
```

Comment:
Maybe not now.

Reply:
Then pls add a TODO here.

Reply:
I don't think we will or are suggested by PM to expose it as a config .

Reply:
If so, you should add some necessary comments here.

Reply:
We need some comments about this magic number.

Reply:
I think the trigger of clean tombstones for lock cf should consider several factors:
1) tombstones occupy how much memory ratio in the skiplist. (if the total memory quota of in-memory engine is 200MB, does 16MB is a good one? How about the total memory quota is 10GB?)
2) Last clean time. In the case of high throughput insertion case, there are a lot of put and delete for the lock cf. For example, all data is embedded in the lock cf's short value part, and the write throughput is 100MB/s, it means every second there would be around 6 times clean lock tombstone tasks and each task will scan the whole lock cf, is the scan efficient?
3) the lock cf has 15MB tombstones and rarely has new write to trigger the clean tombstones jobs, there are some read requests will slow because of these tombstones, should we clean it like "tikv periodically compact lock cf" does?

Reply:
> I think the trigger of clean tombstones for lock cf should consider several factors:
> 
> 1. tombstones occupy how much memory ratio in the skiplist. (if the total memory quota of in-memory engine is 200MB, does 16MB is a good one? How about the total memory quota is 10GB?)
> 2. Last clean time. In the case of high throughput insertion case, there are a lot of put and delete for the lock cf. For example, all data is embedded in the lock cf's short value part, and the write throughput is 100MB/s, it means every second there would be around 6 times clean lock tombstone tasks and each task will scan the whole lock cf, is the scan efficient?
> 3. the lock cf has 15MB tombstones and rarely has new write to trigger the clean tombstones jobs, there are some read requests will slow because of these tombstones, should we clean it like "tikv periodically compact lock cf" does?

Thanks for giving these considerations, there's my thoughts:
1. As a complete write covers put lock and delete lock, every cleanup task will physically deletes majority of the contents in the lock cf. So, it's expected there will not be much bytes in the lock cf. (According to my TPCC test, over 90% of the contents will be deleted in a single cleanup task).
2. The write thoughput 100MB/s is for the whole range in the tikv whereas only a small fraction of the ranges will be cached in the memory and only the cached parts will count the bytes. However, if some regions are pretty hot, the write throughput can still be high. I think we can leave this as an optimization if it really hurt the performance.
3. Good catch, I will leave this as an optimization if we are troubled by this situation.